bootstrap/bootstrap.go

```diff
@@ -173,8 +192,89 @@ func (b *Bootstrap) Cancel() error {
 	return nil
 }
 
+// extractTraceCtx pulls encoded distributed tracing information from the env vars.
+// Note: This should match the injectTraceCtx code in shell.
+func (b *Bootstrap) extractTraceCtx() opentracing.SpanContext {
+	sctx, err := opentracing.GlobalTracer().Extract(opentracing.TextMap, b.shell.Env.ToMap())
+	if err != nil {
+		// Return nil so a new span will be created
+		return nil
+	} else {
+		return sctx
+	}
+}
+
+// stopper lets us abstract the tracer wrap up code so we can plug in different tracing
+// library implementations that are opentracing compatible. Opentracing itself
+// doesn't have a Stop function on its Tracer interface.
+type stopper func()
+
+// startTracing sets up tracing based on the config values. It uses opentracing as an
+// abstraction so the agent can support multiple libraries if needbe.
+func (b *Bootstrap) startTracing(ctx context.Context) (opentracing.Span, context.Context, stopper) {
+	// Newer versions of the tracing libs print out diagnostic info which spams the
+	// Buildkite agent logs. Disable it by default unless it's been explicitly set.
+	if _, has := os.LookupEnv("DD_TRACE_STARTUP_LOGS"); !has {
+		os.Setenv("DD_TRACE_STARTUP_LOGS", "false")
+	}
+
+	buildID, _ := b.shell.Env.Get("BUILDKITE_BUILD_ID")
+	source, _ := b.shell.Env.Get("BUILDKITE_SOURCE")
+	label, hasLabel := b.shell.Env.Get("BUILDKITE_LABEL")
+	if !hasLabel {
+		label = "job"
+	}
+
+	// Set specific tracing library here. Everything else should be using opentracing.
+	// Use a constant sampler - CI runs aren't high traffic.
```

Comment:
It's also worth mentioning that because of how the datadog agent works, this still won't guarantee all spans are sent to APM. Datadog uses a priority-based sampler that rate limits based on the response of the agent after it pushes a batch in case the agent is struggling to keep up.

If you actually want to make sure that it it is kept you need to add
```go
tracer.WithGlobalTag(ext.SamplingPriority, ext.PriorityUserKeep)
```